feat: Added statusCodesToTreatAsSuccess (additive success set) to NewBatchWithFailedRequests and Preserving RequestId

jeppesc11 opened the pull request:
### Changes proposed in this pull request
- Preserving RequestId while using `NewBatchWithFailedRequests `
- Added **statusCodesToTreatAsSuccess** (additive success set) to `NewBatchWithFailedRequests`
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/pull/980)

jeppesc11 committed:
Preserving RequestId for NewBatchWithFailedRequests

jeppesc11 committed:
feat: Added statusCodesToTreatAsSuccess (additive success set) to NewBatchWithFailedRequests

jeppesc11 committed:
Added more test cases NewBatchWithFailedRequests

jeppesc11 commented:
@microsoft-github-policy-service agree

jeppesc11 commented:
@ramsessanchez I can see that I should have explained why - apologies.

The reason is that, for example, with “Get file content” you don’t get an ID back that you can use to correlate with your requests -only the batch request ID. So if some number of the 20 calls in the batch fail and I then retry, I no longer have a way to correlate request and response.

I have previously written a longer blog post about this exact issue.
- https://jeppe-spanggaard.dk/blogs/graph-batching-file-content-mapping/
- https://jeppe-spanggaard.dk/blogs/graph-batch-smart-retry/

The reason for statusCodesToTreatAsSuccess is, again, that with “Get File Content” about 9 out of 10 times I get a response code indicating Redirect or Found. That means these will, by default, be retried - even though it’s a valid status code for this scenario. The same applies to 404.

gavinbarron commented on src/Microsoft.Graph.Core/Requests/Content/BatchRequestContentCollection.cs:
```diff
@@ -183,10 +183,69 @@ public BatchRequestContentCollection NewBatchWithFailedRequests(Dictionary<strin
+
+            HashSet<HttpStatusCode> successSet = statusCodesToTreatAsSuccess != null
+                ? [.. statusCodesToTreatAsSuccess]
+                : null;
```
Why do we have the conversion from IEnumerable to HashSet inside the method? IMO we'd be better served requiring that the HashSet be passed in from the caller.

gavinbarron commented on src/Microsoft.Graph.Core/Requests/Content/BatchRequestContentCollection.cs:
```diff
@@ -183,10 +183,69 @@ public BatchRequestContentCollection NewBatchWithFailedRequests(Dictionary<strin
+                }
+
+                return false;
+            }
```
```suggestion
            bool IsSuccess(HttpStatusCode code) => 
                BatchResponseContent.IsSuccessStatusCode(code) 
                || successSet?.Contains(code) ?? false;
```

jeppesc11 reviewed (commented):
Looks fine with the changes made.

jeppesc11 reviewed (commented):
Looks fine

jeppesc11 committed:
refactor: take HashSet for statusCodesToTreatAsSuccess per review

Address review feedback:
- NewBatchWithFailedRequests now requires the caller to pass a
  HashSet<HttpStatusCode>, removing the IEnumerable-to-HashSet
  conversion inside the method.
- Collapse the IsSuccess local function into an expression body.
  Parentheses around the null-coalescing operand are required:
  without them `a || set?.Contains(code) ?? false` parses as
  `(a || bool?) ?? false` and fails to compile with CS0019.
- The single-argument overload now delegates to the two-argument one
  so the RequestId-preserving step construction lives in one place.
